aten/src/ATen/detail/UniqueVoidPtr.h

```diff
@@ -0,0 +1,79 @@
+#include <memory>
+
+#include <ATen/ATenGeneral.h>
+
+namespace at {
+
+using DeleterFnPtr = void(*)(void*);
+
+namespace detail {
+
+// Does not delete anything
+AT_API void deleteNothing(void*);
+
+// A detail::UniqueVoidPtr is an owning smart pointer like unique_ptr, but
+// with three major differences:
```

Comment:
it owning a context seems fairly different too -- no?  `release_context` doesn't make sense without explaining the context model.  (I think you want to just explain this in 2).

Reply:
new text

```
//    2) It is specialized for a function pointer deleter
//       void(void* ctx); i.e., the deleter doesn't take a
//       reference to the data, just to a context pointer
//       (erased as void*).  In fact, internally, this pointer
//       is implemented as having an owning reference to
//       context, and a non-owning reference to data; this is why
//       you release_context(), not release() (the conventional
//       API for release() wouldn't give you enough information
//       to properly dispose of the object later.)
//

```

Reply:
looks good!